### PR TITLE
aws lambda dead letter queue

### PR DESCRIPTION
**Pull request by muzzamilinovaqo**

### CODE REVIEW

File: plugins/aws/lambda/lambdaDeadLetterQueue.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+                    `Unable to query for Lambda functions: ${helpers.addError(listFunctions)}`, region);
+                return rcb();
+            }
+
+            if (!listFunctions.data.length) {
+                helpers.addResult(results, 0, 'No Lambda functions found', region);
+                return rcb();
+            }
+
+            for (var lambdaFunc of listFunctions.data) {
```
Comment (fatima99s) [inline plugins/aws/lambda/lambdaDeadLetterQueue.js:38]: update the same as Lambda Code Signing Enabled plugin

File: plugins/aws/lambda/lambdaDeadLetterQueue.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+                if (!lambdaFunc.FunctionName) continue;
+                var resource = lambdaFunc.FunctionName;
+                
+                var functionConfig = helpers.addSource(cache, source, ['lambda', 'getFunctionConfiguration', region, resource]);
+         
+                if (functionConfig && functionConfig.data && functionConfig.data.DeadLetterConfig && functionConfig.data.DeadLetterConfig.TargetArn) {
+                    helpers.addResult(results, 0, 'Lambda function has Dead Letter Queue configured', region, resource);
```
Comment (fatima99s) [inline plugins/aws/lambda/lambdaDeadLetterQueue.js:45]:
```suggestion
                    helpers.addResult(results, 0, 'Lambda function has dead letter queue configured', region, resource);
```

File: plugins/aws/lambda/lambdaDeadLetterQueue.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+                } else {
+                    helpers.addResult(results, 2, 'Lambda function does not have Dead Letter Queue configured', region, resource);
```
Comment (fatima99s) [inline plugins/aws/lambda/lambdaDeadLetterQueue.js:47]:
```suggestion
                    helpers.addResult(results, 2, 'Lambda function does not have dead letter queue configured', region, resource);
```

**Review by fatima99s — CHANGES_REQUESTED** [review]

File: exports.js [hunk 4]
```diff
@@ -1621,10 +1622,9 @@ module.exports = {
         'logServiceEnabled'             : require(__dirname + '/plugins/alibaba/ack/logServiceEnabled.js'),
         'cloudMonitorEnabled'           : require(__dirname + '/plugins/alibaba/ack/cloudMonitorEnabled.js'),
```
Comment (fatima99s) [inline exports.js:1624]:
```suggestion


```

File: exports.js [hunk 5]
```diff
@@ -1621,10 +1622,9 @@ module.exports = {
...
 
-
         'securityCenterEdition'         : require(__dirname + '/plugins/alibaba/securitycenter/securityCenterEdition.js'),
         'securityAgentInstalled'        : require(__dirname + '/plugins/alibaba/securitycenter/securityAgentInstalled.js'),
         'securityNotificationsEnabled'  : require(__dirname + '/plugins/alibaba/securitycenter/securityNotificationsEnabled.js'),
         'vulnerabilityScanEnabled'      : require(__dirname + '/plugins/alibaba/securitycenter/vulnerabilityScanEnabled.js')
     }
-};
+};
```
Comment (fatima99s) [inline exports.js:1630]: remove extra line

File: plugins/aws/lambda/lambdaDeadLetterQueue.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'Lambda Dead Letter Queue',
+    category: 'Lambda',
+    domain: 'Serverless',
+    severity: 'Low',
+    description: 'Ensure that AWS Lambda functions are configured to use a Dead Letter Queue.',
+    more_info: 'Dead Letter Queues (DLQs) for your Amazon Lambda functions can make your serverless application more resilient by capturing and storing unprocessed events from asynchronous invocations for further analysis or reprocessing.',
+    link: 'https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-property-function-deadletterqueue.html',
```
Comment (fatima99s) [inline plugins/aws/lambda/lambdaDeadLetterQueue.js:11]: https://docs.aws.amazon.com/lambda/latest/dg/invocation-async.html#invocation-dlq

File: plugins/aws/lambda/lambdaDeadLetterQueue.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'Lambda Dead Letter Queue',
+    category: 'Lambda',
+    domain: 'Serverless',
+    severity: 'Low',
+    description: 'Ensure that AWS Lambda functions are configured to use a Dead Letter Queue.',
+    more_info: 'Dead Letter Queues (DLQs) for your Amazon Lambda functions can make your serverless application more resilient by capturing and storing unprocessed events from asynchronous invocations for further analysis or reprocessing.',
```
Comment (fatima99s) [inline plugins/aws/lambda/lambdaDeadLetterQueue.js:10]:
```suggestion
    more_info: 'Dead Letter Queues (DLQs) are used to store discarded events for further processing, providing a version-specific configuration locked in at publishing, offering a reliable fallback mechanism for failed event processing. They are essential for preserving failed events, aiding in debugging and analysis of processing errors.',
```

File: plugins/aws/lambda/lambdaDeadLetterQueue.js [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+            for (var lambdaFunc of listFunctions.data) {
+
+                if (!lambdaFunc.FunctionArn) continue;
+                var resource = lambdaFunc.FunctionArn;
+                
+                var functionConfig = helpers.addSource(cache, source, ['lambda', 'getFunctionConfiguration', region, lambdaFunc.FunctionName]);
+
+                if (!functionConfig || functionConfig.err || !functionConfig.data) {
+                    helpers.addResult(results, 3,
+                        `Unable to query for Lambda function Config: ${helpers.addError(functionConfig)}`, region, resource);
```
Comment (fatima99s) [inline plugins/aws/lambda/lambdaDeadLetterQueue.js:47]:
```suggestion
                        `Unable to query for Lambda function config: ${helpers.addError(functionConfig)}`, region, resource);
```

**Review by fatima99s — CHANGES_REQUESTED** [review]

File: plugins/aws/lambda/lambdaDeadLetterQueue.js [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+module.exports = {
+    title: 'Lambda Dead Letter Queue',
+    category: 'Lambda',
+    domain: 'Serverless',
+    severity: 'Low',
+    description: 'Ensure that AWS Lambda functions are configured to use a Dead Letter Queue.',
+    more_info: 'Dead Letter Queues (DLQs) for your Amazon Lambda functions can make your serverless application more resilient by capturing and storing unprocessed events from asynchronous invocations for further analysis or reprocessing.',
+    link: 'https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-property-function-deadletterqueue.html',
+    recommended_action: 'Modify Lambda function configurations and enable dead letter queue',
```
Comment (fatima99s) [inline plugins/aws/lambda/lambdaDeadLetterQueue.js:12]:
```suggestion
    recommended_action: 'Modify Lambda function configurations and enable dead letter queue.',
```

File: plugins/aws/lambda/lambdaDeadLetterQueue.js [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+                if (!lambdaFunc.FunctionArn) continue;
+                var resource = lambdaFunc.FunctionArn;
+                
+                var functionConfig = helpers.addSource(cache, source, ['lambda', 'getFunctionConfiguration', region, lambdaFunc.FunctionName]);
+
+                if (!functionConfig || functionConfig.err || !functionConfig.data) {
+                    helpers.addResult(results, 3,
+                        `Unable to query for Lambda function Config: ${helpers.addError(functionConfig)}`, region, resource);
+                    return rcb();
```
Comment (fatima99s) [inline plugins/aws/lambda/lambdaDeadLetterQueue.js:48]: use contine instead of rcb()

**Review by fatima99s — CHANGES_REQUESTED** [review]

File: plugins/aws/lambda/lambdaDeadLetterQueue.js [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'Lambda Dead Letter Queue',
+    category: 'Lambda',
+    domain: 'Serverless',
+    severity: 'Low',
+    description: 'Ensure that AWS Lambda functions are configured to use a Dead Letter Queue.',
+    more_info: 'Configuring Dead Letter Queues (DLQs) with lambda functions helps to store discarded events and isolate messages that can't be processed correctly. It helps to configure an alarm for any messages moved to a dead-letter queue. They are essential in preserving failed events,debugging and determining why their processing didn't succeed. ',
+    link: 'https://docs.aws.amazon.com/lambda/latest/dg/invocation-async.html#invocation-dlq',
```
Comment (alphadev4) [inline plugins/aws/lambda/lambdaDeadLetterQueue.js:11]:
```suggestion
    more_info: 'Configuring Dead Letter Queues (DLQs) with lambda functions helps to store discarded events and isolate messages that can\'t be processed correctly. It helps to configure an alarm for any messages moved to a dead-letter queue. They are essential in preserving failed events,debugging and determining why their processing didn\'t succeed. ',
    link: 'https://docs.aws.amazon.com/lambda/latest/dg/invocation-async.html#invocation-dlq',
```

File: plugins/aws/lambda/lambdaDeadLetterQueue.js [hunk 12]
```diff
@@ -0,0 +1,66 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'Lambda Dead Letter Queue',
+    category: 'Lambda',
+    domain: 'Serverless',
+    severity: 'Low',
+    description: 'Ensure that AWS Lambda functions are configured to use a Dead Letter Queue.',
+    more_info: 'Configuring Dead Letter Queues (DLQs) with lambda functions helps to store discarded events and isolate messages that can\'t be processed correctly. It helps to configure an alarm for any messages moved to a dead-letter queue. They are essential in preserving failed events,debugging and determining why their processing didn\'t succeed. ',
```
Comment (alphadev4) [inline plugins/aws/lambda/lambdaDeadLetterQueue.js:10]:
```suggestion
    more_info: 'Configuring Dead Letter Queues (DLQs) with lambda functions helps to store discarded events and isolate messages that can\'t be processed correctly. It helps to configure an alarm for any messages moved to a dead-letter queue. They are essential in preserving failed events,debugging and determining why their processing didn\'t succeed.',
```

**Review by alphadev4 — APPROVED** [review]

**Review by alphadev4 — APPROVED** [review]

**Pull request merged**